src/sim-host/protocol/socket.js

```diff
@@ -8,77 +8,93 @@ var registerOnInitialize = false;
-        if (!data) {
-            throw 'Exec called on simulation host without exec info';
-        }
-
-        var index = data.index;
-        if (typeof index !== 'number') {
-            throw 'Exec called on simulation host without an index specified';
-        }
-
-        var success = data.hasSuccess ? getSuccess(index) : null;
-        var failure = data.hasFail ? getFailure(index) : null;
 
-        var service = data.service;
-        if (!service) {
-            throw 'Exec called on simulation host without a service specified';
-        }
+    function getFailure(index) {
+        return function (error) {
+            console.log('Failure callback for index: ' + index + '; error: ' + error);
+            var data = { index: index, error: error };
+            socket.emit('exec-failure', data);
+        };
+    }
 
-        var action = data.action;
-        if (!action) {
-            throw 'Exec called on simulation host without an action specified';
-        }
+    function registerSimHost() {
+        socket.emit('register-simulation-host');
+    }
 
-        console.log('Exec ' + service + '.' + action + ' (index: ' + index + ')');
-
-        var handler = pluginHandlers[service] && pluginHandlers[service][action];
-        var telemetryProps = { service: service, action: action };
-        if (!handler) {
-            telemetryProps.handled = 'none';
-            handler = pluginHandlers['*']['*'];
-            handler(success, failure, service, action, data.args);
-        } else {
-            telemetryProps.handled = 'sim-host';
-            handler(success, failure, data.args);
+    Object.defineProperty(module.exports, 'socket', {
+        get: function () {
+            return socket; // Will be undefined if called before initialize().
         }
+    });
 
-        telemetry.sendClientTelemetry('exec', telemetryProps);
+    module.exports.initialize = function (pluginHandlers, services) {
+        serviceToPluginMap = services;
+        socket = io();
+
+        socket.on('init-telemetry', function () {
+            telemetry.init(socket);
+        });
+
+        socket.on('refresh', function () {
+            document.location.reload(true);
+        });
+
+        socket.on('app-plugin-list', function () {
+            // TODO: process the list of plugins
+            socket.emit('start');
+        });
+
+        socket.once('init', function () {
+            socket.on('exec', function (data) {
+                var index;
+
+                if (!data) {
+                    throw 'Exec called on simulation host without exec info';
+                }
+
+                index = data.index;
+                if (typeof index !== 'number') {
+                    throw 'Exec called on simulation host without an index specified';
+                }
+
+                var success = data.hasSuccess ? getSuccess(index) : null;
+            var failure = data.hasFail ? getFailure(index) : null;
```

Comment:
I need to fix linting to include some formatting checks... Half of the file has been indented wrongly.